### ISSUE TITLE
Tracking issue for `#[doc(cfg(...))]`

### ISSUE THREAD

**Issue by kennytm** [rust-lang/rust#43781]
This is a tracking issue for the `#[doc(cfg(...))]` attribute introduced in #43348.

**Steps:**

- [x] Implement
- [ ] Adjust documentation ([see instructions on forge][doc-guide])
- [ ] Stabilization PR ([see instructions on forge][stabilization-guide])

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

(cc #1998)

**Comment by sfackler** [comment]
`#[cfg(rustdoc)]` is also gated on this issue but seems distinct (and less risky). Could we FCP that portion in particular?

**Comment by Nemo157** [comment]
I was just trying this out for documenting crate features, it "works" but if this is a potential usecase it would be nice to special case the rendering for it:

---

<img width="940" alt="Screenshot 2019-07-04 at 12 35 06" src="https://user-images.githubusercontent.com/81079/60660487-2b2a1680-9e58-11e9-9b28-163c9814a14e.png">

---

<img width="454" alt="Screenshot 2019-07-04 at 12 35 27" src="https://user-images.githubusercontent.com/81079/60660526-3f6e1380-9e58-11e9-89c6-dd93088dd374.png">

**Comment by Nemo157** [comment]
There is also the issue that it repeats every feature on every item in a page:

<img width="602" alt="Screenshot 2019-07-04 at 14 17 45" src="https://user-images.githubusercontent.com/81079/60666094-7fd48e00-9e66-11e9-9440-f3a5c249e871.png">

When I last attempted to do something about rendering features I found it much more useful to separately keep track of "all required features" to render at the top of the items page and "newly introduced features" to render on the sub-items on the page, so you don't get this distracting repetition on every item.

**Comment by dtolnay** [comment]
We tried out this feature in Syn (https://github.com/dtolnay/syn/pull/734) and decided against using it yet.

<br>

## What I am happy with

I like how the message turns out at the top of the doc page of a single type or function. 

<p align="center"><img src="https://user-images.githubusercontent.com/1940490/71430597-cd55e900-269a-11ea-94df-638b8289a0db.png" width="80%"></p>

We had previously displayed this information using an italicized note, which was less noticeable.

<p align="center"><img src="https://user-images.githubusercontent.com/1940490/71430661-2c1b6280-269b-11ea-9ec4-5f4071f7fd53.png" width="80%"></p>

<br>

## What I am not happy with

Our index page becomes extremely noisy. I wish there were a way to not show all of these in our case. It is enough to have this information on the type's individual page. Cfg combinations are not among the most important information to show on the index page.

<p align="center"><img src="https://user-images.githubusercontent.com/1940490/71430398-8c110980-2699-11ea-81eb-457fa0c25e83.png" width="80%"></p>

Also inheriting the same note onto every public field seems unnecessary in our use case.

<p align="center"><img src="https://user-images.githubusercontent.com/1940490/71430401-8fa49080-2699-11ea-95bd-a37397faeae5.png" width="80%"></p>

**Comment by Lokathor** [comment]
The links in the opening post have gone dead.

**Comment by jhpratt** [comment]
What's the status of this? I've been using it on the time crate for as long as I can remember, and a number of other crates have been doing so as well.

**Comment by Nemo157** [comment]
For searchability: this is feature `doc_cfg`.

**Comment by Nemo157** [comment]
I changed the rendering of `feature="foo"` cfgs in #75330, which vastly simplifies the display on module index pages like shown in the `syn` screenshot above. I have also just opened #77672 to address the other point @dtolnay had, that there is no need to repeat the exact same `cfg` rendering over and over when it is already implied by context.

Other than those changes, there are bugs around trait implementation handling such as #68100, I want to try and create an exhaustive test covering trait implementations once #77672 is done and fix their handling.

After that, I feel like this would be ready for stabilization, it's had quite thorough usage on docs.rs already, so when rendering is all fixed we can rebuild the documentation for some large crates like `tokio` that use it and check whether there's other edgecase bugs remaining.

**Comment by ArniDagur** [comment]
I'm wondering: Would it not make sense to generate `doc_cfg` hints automatically for any `#[cfg(...)]` items, without also needing to type `#[doc(cfg(...))]`?

So the following
```
#[cfg(all(target_feature = "avx", target_feature = "avx2"))]
#[doc(cfg(all(target_feature = "avx", target_feature = "avx2")))]
pub mod avx;
```
could just be
```
#[cfg(all(target_feature = "avx", target_feature = "avx2"))]
pub mod avx;
```
Otherwise, we're duplicating information unnecessarily, right?

**Comment by Lokathor** [comment]
...pretty much

**Comment by Nemo157** [comment]
In the general case yes, but there are quite a few crates using things like internal `cfg`'s generated by their build.rs for compiler version detection that would not want to show those `cfg` to their users.

**Comment by jyn514** [comment]
What about the opposite - having `#[doc(cfg(x))]` also imply `#[cfg(x)]`? I can't imagine a scenario where you wouldn't want to have that.

**Comment by jyn514** [comment]
Not sure how hard that would be to implement, though - maybe @petrochenkov would know?

**Comment by est31** [comment]
> I'm wondering: Would it not make sense to generate doc_cfg hints automatically for any #[cfg(...)] items, without also needing to type #[doc(cfg(...))]?

Issue with that is if you want to implement the same function on different platforms in different ways. It shouldn't show up as gated on the current platform, but as not gated at all... or if it's only implemented on a set of platforms, then as that list.

So if there is such an implication, it should be opt in, eg via an empty `#[doc(cfg)]` tag. Maybe one can have a feature to make it opt out for a region of code, or an entire crate, but that requires manual review of the code, so can't be the default.

> having #[doc(cfg(x))] also imply #[cfg(x)]?

IMO it's strange if `#[doc(...)]` influences normal code.

**Comment by dtolnay** [comment]
This is mostly looking good for syn. Filed one bug regarding the rendering of doc cfg on impls of empty traits: #79279.

**Comment by jhpratt** [comment]
One situation where it would be nice to have doc_cfg but doesn't currently seem to be feasible is on derive-generated implementations. Even if the provider of the derive wanted to, they'd have to manually accept an equivalent attribute to doc_cfg, only to pass it on.

Not something worth holding up stabilization over, but just something I noticed when adding in more attributes in some of my code.

**Comment by aobatact** [comment]
Can't we have something like a feature-selector, a ui for selecting which feature to show?

**Comment by jyn514** [comment]
I think https://github.com/rust-lang/rust/issues/84437 needs to be fixed before stabilizing this.

> Can't we have something like a feature-selector, a ui for selecting which feature to show?

@aobatact the more features we add to `doc(cfg)`, the long it will be before it's stabilized. I would rather stabilize an MVP and then we can add features later.

**Comment by nagisa** [comment]
There are some issues around how the `#[doc(cfg)]` annotations propagate from module to the items contained within, IIRC. It does only work in some circumstances (though I don't remember which ones, so I can't file a bug…)

That said, I don't think it needs to block the stabilization of the attribute itself – I don't believe the output of the `rustdoc` falls under the stability guarantees, so we can always fix that later.

**Comment by jyn514** [comment]
> One situation where it would be nice to have doc_cfg but doesn't currently seem to be feasible is on derive-generated implementations. Even if the provider of the derive wanted to, they'd have to manually accept an equivalent attribute to doc_cfg, only to pass it on.

Do you want to show the cfg message on *both* the type and the trait? I'm having trouble understanding the use-case - when would that be necessary?

**Comment by nagisa** [comment]
So you can put `#[doc(cfg)]` on a module and it will show up on the types within (or at least used to), but only in _some_ cases. Helps when you have `#[doc(cfg(...))] #[cfg(any(docs, ...)] mod a_ton_of_optional_functionality;` and want to avoid having to annotate with `#[doc(cfg)]` everything within the module.

**Comment by jhpratt** [comment]
I can only speak anecdotally, but I haven't run into any issues with this feature aside from the aforementioned derived traits behind a `cfg` gate (which shouldn't be a blocker imo).

**Comment by jplatte** [comment]
I actually recently saw a case of the doc-cfg hint is rendered on some items when no `doc(cfg())` attribute is applied to those items or any of their parents.

The items in question are all of the methods on this type: https://docs.rs/ruma/0.0.3/ruma/serde/struct.Raw.html
The `doc(cfg())` attribute being rendered does exist in this crate, but only in one place: https://docs.rs/ruma/0.0.3/src/ruma/lib.rs.html#103

**Comment by chorman0773** [comment]
One thing that may be nice would be the ability to provide user-defined text for the "x". I use a number of special features like `thread_impl="c11"`. While these typically don't appear in the public api, having "This is only supported on thread implementation c11" in internal documentation, rather than "This is only supported on thread_impl="c11"" might be nice. 
Not necessarily a blocking feature, though. The way I'd imagine this would be an optional second argument that gives a user-defined representation for the cfg key. In the above case, you'd get `#[doc(cfg(thread_impl="c11","thread implementation"))]` or something similar.

**Comment by eggyal** [comment]
> Do you want to show the cfg message on _both_ the type and the impl? I'm having trouble understanding the use-case - when would that be necessary?

I just hit a case where I'd like the cfg message on the derived impl but *not* the type (the derive attribute itself is applied in a `cfg_attr` attribute).  In particular, it's a `no_std` library that optionally derives some std-only traits if the "std" feature is enabled.  With `doc_cfg`, all my explicit impls of std-only traits are correctly documented as such; but I can't similarly annotate derived ones.

**Comment by eggyal** [comment]
> We're planning to make `doc(cfg)` enabled by default for any `#[cfg]` attribute, would that solve your issue?

It certainly would!

**Comment by kennytm** [comment]
#89596 made `#[doc(cfg)]` implicit when `#[feature(doc_cfg)]` is enabled. It also added `#[doc(cfg_hide(...))]`. So by "stabilizing" I guess you should specify which feature(s) to be stabilized:

* the render `#[cfg]` as `#[doc(cfg)]` behavior?
* the `#[doc(cfg)]` attribute?
* the `#[doc(cfg_hide)]` attribute?

**Comment by daxpedda** [comment]
Currently `#[cfg(any(feature = "danger", test))]` shows up as:
<img width="392" alt="grafik" src="https://user-images.githubusercontent.com/1645124/145498862-dd978a31-3e11-4429-abc4-2c82094ae608.png">

I know `doc_cfg` allows to overwrite, but whats the purpose of showing the `test` tag in documentation?
Is there already an opinion on this matter that I missed?

**Comment by jyn514** [comment]
Happy to take a PR for that :) look in src/librustdoc for where it checks whether `cfg_hide` is active and add `test` to the default set.

**Comment by daxpedda** [comment]
I looked into it, it seems to me that it's a bit more complicated. `cfg_hide` only filters out exact matches, so filtering `test` won't filter test out of `any(feature = "x", test)`.

**Comment by daxpedda** [comment]
After contemplating a couple of solutions, I was wondering if we shouldn't fix this behavior on a different level. Currently `#![doc(cfg_hide(feature = "x"))]` won't hide `x` in `#[cfg(any(feature = "x", feature = "y"))]` or `#[cfg(all(feature = "x", feature = "y"))]`. `cfg_hide` only works on exact matches, does that make sense?

The most straightforward way to me to implement this is to transform `Attribute`s after parsing them, removing anything that should be hidden. This can include `test` of course. But if we don't want to change `cfg_hide`, we could also do this for `test` only.

Any guidance would be appreciated.

**Comment by Nemo157** [comment]
@daxpedda since it looks nontrivial, could you open a separate issue to discuss it on?

**Comment by WaffleLapkin** [comment]
> #89596 made #[doc(cfg)] implicit when #[feature(doc_cfg)] is enabled.

I'm confused. I thought that it worked before, but now it somehow doesn't? Was this behaviour reversed?
```rust
// lib.rs
#![feature(doc_cfg)]

#[cfg(feature = "a")]
pub struct What;
```
```toml
# Cargo.toml
[package]
name = "rustdoc_no_doccfg"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]

[features]
a = []
```
```shell
; cargo +nightly doc --all-features --open
```
![screenshot of rustdoc output showing that feature is not shown](https://user-images.githubusercontent.com/38225716/147862609-bf44cd1b-9083-482e-8b2e-4848d1ca4949.png)
![another screenshot of rustdoc output showing that feature is not shown](https://user-images.githubusercontent.com/38225716/147862645-25e00164-023e-4b81-9c7d-d3524e9a8584.png)
```shell
; cargo +nightly --version
cargo 1.59.0-nightly (fcef61230 2021-12-17)
; rustdoc +nightly --version
rustdoc 1.59.0-nightly (cfa3fe5af 2021-12-31)
```

(adding `#[doc(cfg(...))]` fixes this, but it seemed like it's not required anymore)

**Comment by WaffleLapkin** [comment]
Ah, so since #90502 the implication of `#[doc(cfg(...))]` by `#[cfg(...)]` needs a different feature, `doc_auto_cfg`. It's weird that this wasn't mentioned in this tracked issue at all.

**Comment by Nemo157** [comment]
https://github.com/rust-lang/rust/pull/90502 split that feature out into a separate `doc_auto_cfg` feature, which is as of yet undocumented (I opened https://github.com/rust-lang/rust/issues/92484 for that).

**Comment by vilgotf** [comment]
This feature currently doesn't document feature gated function arguments:
```rust
fn frobnify(#[cfg(feature = "bar")] _bar: ()) {}
```

**Comment by josephg** [comment]
Whats blocking stabilization here? It seems like the feature is complete, and people are using it. If there's tweaks to how documentation is rendered, surely that could happen after stabilization? Devs can already opt out of this for their crates by not adding `#[doc(...)]` attributes in their code.

Are we just waiting on a write-up being merged into the core rust documentation?

I use stable rust, and I find issues like this very frustrating - where there's functionally complete, working compiler features sitting for years behind a flag thats unavailable in stable because ??. It makes stable feel like rust's forgotten stepchild, rather than the main & recommended version of the compiler.

**Comment by lilyball** [comment]
I agree with @josephg, this functionality feels relatively mature and is used all over the place on docs.rs. At this point doing any kind of backwards-incompatible change would break a ton of crate documentation on docs.rs so it feels almost de-facto stable. And since this is still unstable, locally-built documentation feels incomplete, to the extent that I end up running `RUSTDOCFLAGS='--cfg docsrs' cargo +nightly doc` to build local documentation just for `#[doc(cfg(…))]`.

**Comment by eggyal** [comment]
Perhaps this raises a wider issue re stabilisation: if docs.rs runs on nightly/applies unstable features, then those features are at risk of becoming "de facto stabilised".  Perhaps this was already (or should be) raised/discussed elsewhere, but it seems that perhaps this is a Bad Thing.

**Comment by nagisa** [comment]
`docs.rs` generates documentation once and stores it as is. If it looks good at the time of generation, it will remain that way indefinitely. While changing the way this functionality works today would affect the documentation output if a new version of the crate was published after such a change, it is also pretty clear cut that the crate opts into nightly functionality knowingly by asking docs.rs to use an unstable feature with the specific version of the Rust that docs.rs happens to be using at the time.

I don't see the risk as significant here. With that context I'm also opposed to freezing the current functionality or rushing the stabilization.

---

> Are we just waiting on a write-up being merged into the core rust documentation? I use stable rust, and I find issues like this very frustrating...

A great way to have this particular itch scratched is to contribute some work towards scratching it. One thing that's missing for the stabilization is a [stabilization report](https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#write-a-stabilization-report) which could among other things give an objective overview of the [known issues](https://github.com/rust-lang/rust/labels/F-doc_cfg) and why they don't block the stabilization of this feature. It should be posted as a comment on this issue.

**Comment by GuillaumeGomez** [comment]
https://github.com/rust-lang/rust/pull/89596 was merged a while ago. Is there anything else remaining to be done in here? Otherwise I can open a stabilization PR like https://github.com/rust-lang/rust/pull/79263.

**Comment by jhpratt** [comment]
I'm not aware of any blockers. I have been using this for many months without issue, as have many other large crates.

**Comment by GuillaumeGomez** [comment]
Same on my side but I was thinking that maybe I was missing something.

Do you think we can stabilize it @rust-lang/rustdoc ?

**Comment by est31** [comment]
I'm wondering about this *and* here.
![Screenshot_20220523_173414](https://user-images.githubusercontent.com/8872119/169855464-396cd517-f7e6-4e0c-b7c5-e52c4fa9e225.png)

It's a bit ambiguous what *and* entails. It can be interpreted by users that enabling either of the two features is enough for getting that function.

Maybe "Requires crate features http and cache" would make it clearer?

**Comment by GuillaumeGomez** [comment]
`Requires windows` seems a bit weird but I'm fine with it.

**Comment by najamelan** [comment]
Stabilization of this would be a massive quality of life improvement.

**Comment by jhpratt** [comment]
For clarity, the stabilization is regarding the attribute itself. How it's displayed, like everything else in docs, can change to be improved. So wording isn't a blocker.

**Comment by Nemo157** [comment]
Is the plan to only stabilize `doc(cfg)`, or all three features? If the latter, `feature(doc_auto_cfg)` probably needs to have an explicit opt-in like `#![doc(auto_cfg)]` added so that we don't start just showing cfgs on crates that aren't prepared for it (and maybe a plan to make that default active in a future edition); and some usage reports for `cfg_hide` would be nice.

**Comment by GuillaumeGomez** [comment]
Only `doc(cfg)` iiuc.

**Comment by GuillaumeGomez** [comment]
However I think most people in here are very likely talking about `doc_auto_cfg`. Might be better to clarify beforehand.

**Comment by SpriteOvO** [comment]
Does `cfg_hide` currently not support hide `not()` feature?

```rust
#[cfg(any(all(windows, feature = "winapi"), all(doc, not(doctest))))]
pub use win_debug_sink::*;
```

I need `not(doctest)` here because without it `cargo test` will run doctests in the `win_debug_sink` mod on a platform other than Windows.

In the head of my `lib.rs`:

```rust
#![cfg_attr(
    all(doc, CHANNEL_NIGHTLY), // `CHANNEL_NIGHTLY` is passed by `build.rs`
    feature(doc_auto_cfg, doc_cfg_hide),
    doc(cfg_hide(doc))
)]
```

It generates badge: `Available on Windows and crate feature winapi, or non-doctest only.`

It's good that the `doc` is properly hidden. But if I add a `doc(cfg_hide(doctest))`, the badge just disappeared, instead if I add a `doc(cfg_hide(not(doctest)))`, the badge is back, but it doesn't seem to be doing anything.

---

Or consider another solution, maybe we should remove `doc` and `doctest` in the same way as we did in issue #91740?

BTW, the implementation in #91740 also only removes `test` but not `not(test)`, the latter still makes the badge disappear. Looks like a bug.

**Comment by dns2utf8** [comment]
Is the way with defining a `doc_auto_cfg` feature still supported currently? And in the future?

**Comment by dns2utf8** [comment]
There used to be a feature the the docs.rs renderer enabled, like play.rust-lang.org, so as a crate author I could enable specific features for the respective builds.

Also, I found this example that could solve the same problem: https://github.com/rust-lang/docs.rs/blob/369e1bb0ed5f0b4c14f6ecaa01013094c7d234dd/templates/core/Cargo.toml.example
```
[package.metadata.docs.rs]

# Features to pass to Cargo (default: [])
features = ["feature1", "feature2"]

# Whether to pass `--all-features` to Cargo (default: false)
all-features = true
```

**Comment by GuillaumeGomez** [comment]
I am very lost. This is not linked to that (at least not directly). `doc(cfg(...))` is used to show the `cfg`s independently or the current `cfg`.

**Comment by Kixunil** [comment]
Could we get a small warning triangle or a similar marker with hover at declaration site? Consider this example:

![Screenshot 2022-12-06 at 13-57-54 Foo in foo - Rust](https://user-images.githubusercontent.com/1178779/205921621-8257002a-8822-4e9b-8323-107f1a764572.png)

The conditionality is not immediately visible, the user has to scroll down. For small enums this is OK but large enums that have each variant extensively documented are a problem. If one doesn't scroll down to check if it requires a feature it could result in annoying "why it says it doesn't exist?" confusion.

**Comment by tmccombs** [comment]
Is anyone already working on creating an RFC?

**Comment by GuillaumeGomez** [comment]
Not that I know. I have it in my TODO list.

**Comment by jyn514** [comment]
As a remind, this is a tracking issue for *stabilizing* `doc_cfg`. Bug reports should go in separate issues to avoid pinging the 30+ people watching this issue, unless you think the bug is so severe it should block stabilization. Suggestions for enhancements should also go in separate issues.

---

As far as I can tell, the current status is that the stabilization PR (https://github.com/rust-lang/rust/pull/100883) is blocked on deciding whether `doc_auto_cfg` should be on by default or not (and whether we should let crate authors control whether it's on or just force it on). The discussion for that is in https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/Discussion.20about.20.60doc_cfg*.60.20features.20stabilization/.

**Comment by GuillaumeGomez** [comment]
Thanks for the sum up. We actually decided to go through an RFC that started to be written because too many elements need to be discussed. More information in the next weeks.

**Comment by joseluis** [comment]
Since I see an RFC is being written, I take the opportunity to share my case, since I wanted to use `doc_cfg` to show that a particular function is only available on the absence of a feature. But it seems that's not currently possible.

This does result in: "Available on feature `safe` only". 
```
#[cfg(feature = "safe")]
#[doc(cfg(feature = "safe"))]
pub fn safe_fn() {}
```

But this results in: "Available on non-crate feature `safe` only". Instead of the expected: "Not available on feature `safe`".
```
#[cfg(not(feature = "safe"))]
#[doc(cfg(not(feature = "safe")))]
pub fn unsafe_fn() {}
```

I wonder if this is by design.

**Comment by Nemo157** [comment]
> I wonder if this is by design.

Somewhat, yes. Cargo itself [doesn't support negative features](https://doc.rust-lang.org/nightly/cargo/reference/features.html#feature-unification), so no effort was put into making them render well.

**Comment by nic-hartley** [comment]
Should the original post link to #100883 instead?

**Comment by seancroach** [comment]
I'm running into a rather annoying problem with documenting features.

I am writing currently a `serde` crate with multiple specifications for a specific format. I have a `binary` module that is documented with the following `doc_cfg`: `any(feature = "be", feature = "le", feature = "varint")`. Now, if I document one of the functions as present with just the `be` feature and inline their re-export I get this annoying display:

![image](https://user-images.githubusercontent.com/10006774/232262195-d6300b37-9d39-4248-ab22-e26561eb2160.png)

Of course, I tried with `doc_cfg_hide`, wrestling with the minimal documentation, but nothing seems to work for this. I can *not* inline the re-exports, but purposefully sacrificing usability is presumably a bug or at least something worth mentioning here. When I don't inline it, it looks like so:

![image](https://user-images.githubusercontent.com/10006774/232262239-b350a830-651b-4207-8d01-b86ff40bca59.png)

My one gripe with the first selection is that even the order it's appended to is non-uniform, which bothers me, but that's not the core issue here.

Is there some hack I'm unaware of besides not inlining items in documentation? I haven't seen this case discussed so far. Even if I simplified it to an internal feature (`binary`) it still displays, even if assumed to be enabled in context; either way, extraneous information is applied, and I have *no say* in omitting it.

**Comment by GuillaumeGomez** [comment]
`#[doc(no_inline)]`

**Comment by seancroach** [comment]
I believe `#[doc(no_inline)]` (as described [here](https://doc.rust-lang.org/rustdoc/write-documentation/the-doc-attribute.html#inline-and-no_inline)) is *specifically* the behavior I'm trying to avoid. As you can see, `binary` is a public module.

![image](https://user-images.githubusercontent.com/10006774/232335410-4058de72-7d4c-4e52-80fb-dc130b6a75f1.png)

I understand this is stylistic, but every `serde_*` crate uses `#[doc(inline)]` on their re-exported functions. This is the behavior I want.

The behavior I *don't* want is my inability to omit the automatically appended features. It introduces a ton of noise. Even if I make some "internal" feature named binary, it still (non-uniformly in terms of order) adds itself to the re-exported functions. I've made this an issue here because it seems like not too rare of a use-case that makes documenting features so noisy that it becomes unhelpful at-a-glance. 

In summary, I can either not inline my re-exports, or deal with unhelpful documented features; it seems like a bug that I can't simply do both, even if it means manually overriding the item at its declaration to say *exactly* what I want to document.

**Comment by Nemo157** [comment]
It should certainly be possible to simplify `be and std and (be or le or varint)` to `be and std`, IIRC there are already some simplifications applied to the `cfg`s so this would just require extending those. It'd be good to open a separate issue with a small testcase.

**Comment by est31** [comment]
@Nemo157 see #104991

**Comment by GuillaumeGomez** [comment]
We'll need https://github.com/rust-lang/rust/issues/103300 to be fixed beforehand as well.

**Comment by daxpedda** [comment]
I noticed in https://github.com/gfx-rs/wgpu/pull/4935#discussion_r1436658697 that it would be very nice if we could add custom names to `cfg`s. This is currently done by Rustdoc for known `cfg`s here:
https://github.com/rust-lang/rust/blob/2b1365b34f0d5ee43944c4266a625923a7b312dd/src/librustdoc/clean/cfg.rs#L488-L590

Maybe something like `doc(cfg_rename(foo = "Foo"))`?

**Comment by Kixunil** [comment]
I just found this doesn't appear to support `version(..)` cfg. I currently workaround the lack of version predicate by having a build script detect the rustc version and output a `cfg`. I thought adding `#[cfg_attr(docsrs, doc(cfg(any(feature = "std", version("1.77.0")))))]` would cause it to render properly but it just fails to build with "invalid predicate" message. I tried adding `#![feature(cfg_version)]` without any luck.

**Comment by GuillaumeGomez** [comment]
`version(...)` is an internal rustc attribute, so it's not supposed to be used outside of std.

**Comment by Kixunil** [comment]
It isn't, see #64796

**Comment by GuillaumeGomez** [comment]
This issue is 5 years old and doesn't seem to have been updated much since then. And the only place where the attribute is used is in std libs. However it can be used with a nightly rustc:

```rust
#[cfg(version = "1.70.0")]
fn foo() {}

fn main() {}
```

gives:

```
error[E0658]: `cfg(version)` is experimental and subject to change
 --> src/main.rs:1:7
  |
1 | #[cfg(version = "1.70.0")]
  |       ^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #64796 <https://github.com/rust-lang/rust/issues/64796> for more information

For more information about this error, try `rustc --explain E0658`.
```

But in any case, until there is a clear position on whether or not this attribute should be rustc internal only or not, I don't think there is anything to be done.

**Comment by Kixunil** [comment]
The issue I linked is literally a tracking issue for an accepted RFC. It's not internal, it's supposed to be usable in nightly with syntax `version("1.77.0")`, not `version = "1.77.0"`. It should be processed by rustdoc correctly to give us a view how things might look once both features are stable.

**Comment by jhpratt** [comment]
@GuillaumeGomez frankly I have no idea what you're thinking of, but you're unquestionably confusing two different features.

**Comment by clarfonthey** [comment]
Just poking around: it appears that the `doc_auto_cfg` feature links here, but there's no docs in the unstable book for it and the feature doesn't seem to match what's proposed in the RFC. Is this just not implemented, or is the syntax different and this just isn't documented?

**Comment by GuillaumeGomez** [comment]
The RFC (which is still under discussion so maybe not its final form) proposes something slightly different than what's currently implemented and adds more attributes to give more control.

Currently, to enable this feature, you use `#![doc(auto_cfg)]` and that's it.

**Comment by jmillikin** [comment]
Two years ago on https://github.com/rust-lang/rust/pull/100883 I asked about the possibility of stabilizing `#[doc(cfg(...))` without `#[doc(auto_cfg)]` on the principle that the former seems to be well-understood, while the latter is a more complex feature that has ongoing discussion about how exactly it should work.

At the time, the response was that stabilizing `#[doc(cfg(...))]` by itself wouldn't provide a good user experience, and that there would potentially be ecosystem churn from people adding it all over and then removing it once `#[doc(auto_cfg)]` was ready.

But ... it's been two years, and neither is stabilized. I could have been using `#[doc(cfg(...)]` all this time. It's pretty frustrating to have a ready-to-stabilize feature I want be blocked on a vaguely-related feature that I don't particularly care about.

Can `#[doc(cfg(...))]` please be stabilized? Feel free to put some verbiage in the release notes like "if you're planning to use `#[doc(auto_cfg)]` in the future then please hold on until that arrives" if there's still a concern about ecosystem churn.

**Comment by jhpratt** [comment]
@jmillikin This needs an RFC to be stabilized. That's currently awaiting FCP (it needs 3 more checkboxes): https://github.com/rust-lang/rfcs/pull/3631#issuecomment-2208528823

**Comment by jmillikin** [comment]
I understand why `#[doc(auto_cfg)]` needs an RFC.

Why would `#[doc(cfg(...))]` need an RFC?

**Comment by jhpratt** [comment]
It's a new, non-trivial language feature. They don't get added on a whim; there's a process to follow.

**Comment by theemathas** [comment]
For reference, writing the following at the top of `lib.rs` will make feature-gated stuff documented properly on docs.rs, while still having your program compiling in stable rust:

```rust
#![cfg_attr(docsrs, feature(doc_auto_cfg))]
```

**Comment by jmillikin** [comment]
`#[cfg_attr(docsrs), ...]` doesn't help for people who are trying to run rustdoc as part of their build and/or CI process, so it would be nice to get support for at least `#[doc(cfg(...))]` stabilized.

**Comment by GuillaumeGomez** [comment]
It's all or nothing. RFC is open and waiting for reviews. Sadly there is nothing more we can do about it.

The `cfg_attr` trick also works with features. Nothing prevents you do to run your CI in nightly. It's not the best but until the RFC is merged and the feature implemented (this part should be pretty quick), nothing else can be done.

**Comment by jmillikin** [comment]
I'm sorry, this may be a really dumb question, but why is it "all or nothing"?

I don't understand why `#[doc(cfg(...))]` can't be stabilized as an MVP, independent of `#[doc(auto_cfg)]`. They're both already implemented, and from the perspective of writing a stabilization report and/or updating the reference the former feature seems a lot easier to document (since it's entirely explicit annotations by the author).

**Comment by jhpratt** [comment]
Theoretically it could be done separately, but it wouldn't be done any faster (which seems to be what you truly want). There is _nothing_ that can be done to speed up stabilization (at least that's within the code of conduct).

**Comment by jmillikin** [comment]
> Theoretically it could be done separately, but it wouldn't be done any faster (which seems to be what you truly want). There is _nothing_ that can be done to speed up stabilization (at least that's within the code of conduct).

I've frequently seen subsets of a complex feature stabilize separately, most famously the various parts of `async` stabilizing as the design was iterated on. I've also seen existing RFCs get split up into an MVP and a follow-on with more complexity but better ergonomics.

It could be that there's some aspect of how `#[doc(cfg(...))]` and `#[doc(auto_cfg)]` interact that I'm missing due to lack of `rustc` internals knowledge, but from an outside perspective it's not obvious why `#[doc(cfg(....))]` couldn't be split off and stabilized by itself (e.g. in 1.85 or 1.86), and then `#[doc(auto_cfg))]` can stabilize on its own timeframe (presumably >= 1.90).

**Comment by jhpratt** [comment]
Note that when features are split up it's generally because there are concerns, unresolved questions, or simply too much work to do all at once. None of that applies here. The only blocker (pending concerns being raised) is checkboxes.

**Comment by GuillaumeGomez** [comment]
In addition to what @jhpratt said: because it would need to go through an RFC, which means asking people to write another RFC and yet other people to review it. We have limited bandwidth and most contributors are not paid to do so and we're also facing a huge burn-out issue across the project as well so it wouldn't help in this regard either.

**Comment by madsmtm** [comment]
[I've pinged the people that the FCP is waiting for](https://github.com/rust-lang/rfcs/pull/3631#issuecomment-2437608000), hopefully that can speed up the progress a little bit.

**Comment by GuillaumeGomez** [comment]
1. Please don't.
2. Likely not.

**Comment by madsmtm** [comment]
Apologies then.

**Comment by SpriteOvO** [comment]
> `#[cfg_attr(docsrs), ...]` doesn't help for people who are trying to run rustdoc as part of their build and/or CI process

@jmillikin This is the solution I have been using.

`Cargo.toml`
```toml
[build-dependencies]
rustc_version = "0.4.1"
```

`build.rs`
```rust
use rustc_version::{version_meta, Channel};

fn main() {
    // Set cfg flags depending on release channel
    let channel = match version_meta().unwrap().channel {
        Channel::Stable => "CHANNEL_STABLE",
        Channel::Beta => "CHANNEL_BETA",
        Channel::Nightly => "CHANNEL_NIGHTLY",
        Channel::Dev => "CHANNEL_DEV",
    };
    println!("cargo:rustc-cfg={}", channel)
}
```

`lib.rs`
```rust
#![cfg_attr(all(doc, CHANNEL_NIGHTLY), feature(doc_auto_cfg))]
```

Then run nightly in your CI script, `doc_auto_cfg` will only be enabled when using the nightly toolchain. So this works very well with CI and docs.rs and doesn't affect library users.

**Comment by fjarri** [comment]
The way I have it set up:

`Cargo.toml`:
```
[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--cfg", "docsrs"]
```

`lib.rs`:
```
#![cfg_attr(docsrs, feature(doc_cfg, doc_auto_cfg))]
```

CI (mimicking docs.rs settings):
```
env RUSTDOCFLAGS='--cfg docsrs -D warnings' cargo doc --all-features
```

@jmillikin, does that not solve your problem?

**Comment by joshka** [comment]
I noticed today that feature flags on derived traits don't get annotated and there's no obvious way to add an annotation

e.g.:

```rust
#![feature(doc_auto_cfg)]
#![feature(doc_cfg)]

use std::fmt;

// this doesn't display the feature flag
#[cfg_attr(unix, derive(Debug))]
pub struct Foo;

// this works
#[cfg(unix)]
// #[doc(cfg(unix))] // if not using auto_cfg
impl fmt::Display for Foo {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Foo")
    }
}
```

rendered with cargo-docs-rs:
![Image](https://github.com/user-attachments/assets/f4c8cf20-c784-48a0-a927-3abc7ea078a2)

Is there a way that this should be able to work, or is it a known missing part of this? If the latter, I'd note that it's likely not a blocking one IMO, but a nice to have.

**Comment by jhpratt** [comment]
I'm not sure how that even _could_ work theoretically, given that `#[cfg_attr]` is evaluated first.

**Comment by joshka** [comment]
Perhaps something like the following?:
```rust
#[cfg_attr(unix, derive(Debug))]
#[doc(cfg_impl(unix, Debug))]
struct Foo;
```

What is it about the code generated by the compiler `derive` attribute that makes it not able to be annotated the same way a manual `impl` might be?

**Comment by est31** [comment]
As another method to what @joshka said, one could introduce a `#[doc(cfg_attr(...))]` attribute and make `cfg_attr` expand to it in addition to `#[derive(Debug)]`. It would be ignored by rustdoc, as making sense of it should be job of the derive macros: one would need to add support to all the derive macros to recognize that `#[doc(cfg_attr(...))]` attribute and transform it to a `#[doc(cfg(...))]` on the impl block.

The advantage would be that it's automatic compared to `cfg_impl`, but it would need derive authors to opt into this.

If you don't do such a transformation, like in @joshka 's proposal, rustdoc would need some way to find the right impl block (thanks to generics there can be multiple). Eg it could put the note on all impl blocks that come from expansions (`Span::from_expansion`).

There is also the whole question of what to do in the case the cfg evaluates to false and the derive macro does not run. You can't just create a fake `impl Debug for Foo`, as you need to get the generic params right: it could just as well be a `impl<T: Debug + SomeTraitFromFooTypeDecl> for Foo<T>`. Probably the impl block would just not be shown in that case.

**Comment by jhpratt** [comment]
I thought about this some more...it may be possible if the attribute is stored along with an annotation as to how it was enabled (if not by nature of not needing to be). This is already done to an extent as far as I'm aware, as diagnostics knows about `cfg`ed out items.

**Comment by joshka** [comment]
To be clear, my `cfg_impl` proposal was not a fully formed / thought out one. It was very off the cuff and I expect there's likely many downsides to be explored before jumping on that idea. My knowledge of the implementation necessary for this is fairly limited.

**Comment by GuillaumeGomez** [comment]
It's supposed to work. It's a known issue: https://github.com/rust-lang/rust/issues/103300

The problem is rustc doesn't provide information when expanding items under `cfg_attr`. The best way here would be to add the corresponding `cfg` attribute on each generated item.

**Comment by jhpratt** [comment]
For those only subscribed to the tracking issue, the RFC is currently in FCP: https://github.com/rust-lang/rfcs/pull/3631#issuecomment-2701673067

**Comment by lolbinarycat** [comment]
when this is stabilized, the feature flag should be ignored, as many packages have `#![cfg_attr(docsrs, feature(doc_auto_cfg))]`, and removing this feature all together would mean that docs.rs's continuous rebuilds would break for packages until they added a new release that removed the `cfg_attr` line.

**Comment by kennytm** [comment]
@lolbinarycat docs.rs is building using a [nightly compiler](https://docs.rs/about/builds), so it won't cause [E0554](https://doc.rust-lang.org/stable/error_codes/E0554.html). Use of stabilized feature will only emit a `stable_features` lint e.g.

```
warning: the feature `if_while_or_patterns` has been stable since 1.33.0 and no longer requires an attribute to enable
 --> src/main.rs:1:28
  |
1 | #![feature(range_is_empty, if_while_or_patterns)]
  |                            ^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(stable_features)]` on by default

warning: the feature `range_is_empty` has been stable since 1.47.0 and no longer requires an attribute to enable
 --> src/main.rs:1:12
  |
1 | #![feature(range_is_empty, if_while_or_patterns)]
  |            ^^^^^^^^^^^^^^
```

I don't see how the existing practice is going to break any packages.

**Comment by lolbinarycat** [comment]
ah, alright, i wasn't sure what happens to stabilized features.  i suppose its only removed features that have their identifiers removed without a trace.

**Comment by kennytm** [comment]
it is already using `--cap-lints warn` 5 years ago (rust-lang/docs.rs#659)

**Comment by lolbinarycat** [comment]
looks like everythings already been thought of, then.

**Comment by GuillaumeGomez** [comment]
Good idea! Also linked the implementation PR in the first comment (https://github.com/rust-lang/rust/pull/138907).

**Comment by dhardy** [comment]
Clarification needed: is `#[doc(cfg(not_a_known_cfg))]` valid?

The RFC does not state this explicitly, hence I would assume that it is, and indeed `rustdoc` will output a corresponding annotation.

Attempting this does, however, generate a warning: unexpected `cfg` condition value: `not_a_known_cfg`.

Is this an inadequacy of the specification or is it a bug in the `unexpected_cfgs` lint?

---

Motivation for asking: I have been using this to optionally show otherwise hidden `pub` items with an annotation, for example [here](https://github.com/kas-gui/kas/blob/31f164f206be832f5df75115b1ea1986fc462396/crates/kas-core/src/util.rs#L203). (These annotations don't show on cross-crate re-exports, but that's an [unrelated issue](https://github.com/rust-lang/rust/issues/67295#issuecomment-3019605242) which should be solved once this feature is stabilised.)

**Comment by GuillaumeGomez** [comment]
Same as for `cfg(not_a_konwn_cfg)`. You can do it, and you will get a warning for it. You can add the `cfg` to the list in your `Cargo.toml` file.
